### ISSUE TITLE
Unintelligible console output

### ISSUE THREAD

**Issue by williamboman** [npm/npm#8812]
When running `$ npm install` on `npm@3.1.0` `node@0.12.5` the console output is pretty unintelligible. Attached gif for demo purposes.

![npm-install](https://cloud.githubusercontent.com/assets/6705160/8500571/bb40df24-219d-11e5-85df-0bc622c0e7c5.gif)

Somewhat related to #8761. This issue doesn't only refer to the carriage return/new line issue.

**Comment by iarna** [comment]
Yes, if you change your window size while it's running it will mess it up. Uh... don't do that?

Yeah, we can try to catch SIGWINCH and look at the terminal size again although that's not universally available, so.

**Comment by williamboman** [comment]
@iarna This behaviour is also present when working with split panes, which I do constantly. It's so malfunctioning I have to change the way I work in the terminal. No other package manager makes me have to do that. I really think it should be addressed.

**Comment by iarna** [comment]
Hmm, I use split panes all the time and I've never encountered an issue… something somewhere must be misreporting the terminal size on you. Hmm…

**Comment by williamboman** [comment]
@iarna Oh, no issues with terminal sizing. It's just that I like to work in other panes (often creating new ones) while an installation is ongoing. Listening to SIGWINCH and resizing the progress bar accordingly would fix this.

What's the status on the progress bar? Is it a WIP or is it considered pretty much done?

**Comment by iarna** [comment]
Pretty much done. The module providing the progress bar visuals is here:

https://github.com/iarna/gauge

At some point, I intend to make a `meta-gauge` library that exposes that API but allows plugging w/ a host of different progress indicators. That will probably happen when we start looking at making npm more embedable. (Specifically, one of those views would be to just print out progress updates in a parsable format to allow things calling npm (like Visual Studio) to display their own progress bars based on those updates.)

**Comment by iarna** [comment]
Completion tracking is a separate concern and is isolated to https://github.com/iarna/are-we-there-yet, which has its top level change event trigger gauge `show` calls.

`npmlog` triggers `pulse` calls with every log line.

**Comment by iarna** [comment]
The "listen to column changes" fix is in, though some cruft still shows up, so I'm going to defer calling this fixed until we get that cleaned up.

**Comment by williamboman** [comment]
The "cruft" is pretty much unavoidable afaik. Perfectly acceptable imo.

**Comment by iarna** [comment]
Some of it I think I can fix 😁(Probably not when you shrink of course, but it shouldn't cruft up when you grow.)

**Comment by williamboman** [comment]
Huh, testing it a bit further I agree. Did you have something like `process.stdout.on('resize', redraw)` in mind?

**Comment by iarna** [comment]
Yeah, plus clearing to the end of the line.

**Comment by iarna** [comment]
I ended up going with https://github.com/iarna/gauge/commit/ca15696497ca3100392dfcb6a97705ec4fa9680f which just does a hide/show if we're already showing it. I can grow the window without any cruft.

**Comment by shreeve** [comment]
Random comments about the gauge/bar. When used with npm 3, it takes a huge portion of the horizontal width (maybe 75%), leaving only a small amount for the actual text. It seems it should be the other way around. The user seems to want to know "where they are" in the process, but that can be accomplished with a friendly little bar and more space allotted to the text. As it is now, the text has so little room that it's become sort of useless. Also, the "end caps" seem unnecessary. Other improvements seem great.

**Comment by shreeve** [comment]
Possibly a little cleaner look... a change in the order of the output at the following unicode tweaks:

```javascript
ProgressBar.unicode = {
  startgroup: " ",
  endgroup: " ",
  complete: "█",
  incomplete: "-",
  spinner: "◐◓◑◒",
  subsection: "→"
}
```

![npm-bar](https://cloud.githubusercontent.com/assets/142875/8592288/1c4deeb2-25fc-11e5-8dc1-75e315042a36.gif)

**Comment by iarna** [comment]
I don't believe that unicode is available on windows...

**Comment by williamboman** [comment]
@shreeve I like it!

**Comment by shreeve** [comment]
@iarna - It's more than just the leading spinner... It's a few changes:

* remove the ```startgroup``` and ```endgroup``` characters (they detract from, not enhance, the output)
* replace the ```incomplete``` "shaded block" with a simple dash (it's nice to be able to "count", which is hard when chars run together)
* move the progress bar to the left side, so everything lines up cleanly and is more consistent
* make the progress bar more manageable (ie - 10 chars makes each tick 10%); having a huge bar with fine-grained percentages is too pedantic
* move the "text" portion to the right of the bar, so you have plenty of room to output values instead of squeezing 'em before the bar
* replace the ```spinner``` chars with something that is more consistent (each char the same size, etc.); the original blocks were totally inconsistent

**Comment by iarna** [comment]
@shreeve I know, but that was the only thing that was actually a no-go, so it's what I called out.

The rest requires greater discussion because it comes down to aesthetics and different needs. I want the progress bar to actually move, even on people's absurdly large, 8000 module projects.

**Comment by iarna** [comment]
I originally used characters from the same block as the one's you picked, but Windows commandline limitations are pretty heavy here. In a perfect world I'd use clock emoji for that. (You can have as many characters in a spinner as you like. It just happened that the examples were cycles of 4.)

**Comment by iarna** [comment]
Hey, could you open a new issue with just your suggested changes in it? (I like to separate "I'm having a problem..." from "here's a potential solution" in terms of open issues– makes it easier to manage.)

**Comment by iarna** [comment]
@williamboman You mention the "carriage return/newline issue" in your summary but… it's not clear to me what you're referring to?

**Comment by williamboman** [comment]
@iarna Basically how it would not adjust to window size changes (fixed in https://github.com/iarna/gauge/pull/3).

**Comment by iarna** [comment]
Ok, so my plan now is:

* [x] Landing the improved resize handling (happening 7/17)
* [ ] Add support for configuring the theme & spinner via the npm config, which will make it a lot easier to experiment with different layouts.

**Comment by fresheneesz** [comment]
This seems related: https://github.com/npm/npm/issues/10469

**Comment by shellscape** [comment]
@shreeve holy smokes I love that spinner gif
